### PR TITLE
Add developer guidelines regarding debug/trace logging.

### PR DESCRIPTION
**Pull request by matriv**

### CODE REVIEW

File: devs/docs/logging.rst [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+Therefore, it is advisable, when working on such components (implementing new
+features, refactoring code or maybe just debugging) to try to introduce helpful
+log messages. Those messages will not only help with the current work but will
+be extremely useful for future work on those components.
+
+Such messages should be logged in ``DEBUG`` or ``TRACE`` level depending on how
+fine grain information they provide. To avoid introducing performance overhead
+they must also be "guarded" by checking the enabled level of the logger, e.g.::
+
+  if (logger.isTraceEnabled()) {
```
Comment (mfussenegger) [inline devs/docs/logging.rst]: Might want to note that this should ideally be done only once if the component is initialized, unless it's a singleton.

`isXEnabled` itself is relative expensive.

Also, maybe link to https://www.brandur.org/logfmt as a guideline for how to format the log statements.

Comment (matriv) [reply devs/docs/logging.rst]: Need to rephrase it yes, the statement should be guarded if does some "computation" or has more arguments than the defined methods and we fallback to the vararg method signature.

File: devs/docs/logging.rst [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      logger.trace("My useful log message");
+  }
+
+For components of the distributed execution engine (e.g.:
```
Comment (mxm) [inline devs/docs/logging.rst]: distributed `query` execution

File: devs/docs/logging.rst [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+       public MyClass(Settings settings, ...) {
+           logger = Loggers.getLogger(MyClass.class, settings);
+           ...
+        }
+  ...
+  }
+
+In your log messages try to include also information about the "environment" in
+which the operation takes place. Such information can be for example the
+`nodeOperations`, the `phaseId`, etc.
```
Comment (mxm) [inline devs/docs/logging.rst]: the `padId`, or others.

Comment (matriv) [reply devs/docs/logging.rst]: padId ?

File: devs/docs/logging.rst [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+expensive computation or they use more than 10 arguments, they must also be
+conditional on the current level of the logger, e.g.::
+
+  if (logger.isTraceEnabled()) {
+      // Relatively expensive calculations
+      ...
+      logger.trace("My useful log message");
+  }
+
+``isXXXEnabled`` are also relatively expensive so if possible save it to a
```
Comment (nslater) [inline devs/docs/logging.rst]: These sorts of function calls (\`\`isFooEnabled\`\`) are relatively expensive themselves, so if possible, save the result to a class attribute during object initialisation.

File: devs/docs/logging.rst [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+class attribute during its initialisation to avoid multiple calls to these
+methods.
+
+For components of the distributed execution engine (e.g.:
+``PageDownstreamContext``), additional node information is very useful. To
+achieve that, do not use a ``static`` logger instance, as usual. Instead
```
Comment (nslater) [inline devs/docs/logging.rst]: "Instead,"

File: devs/docs/logging.rst [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+logger instance should become a class attribute which is initialized with the
```
Comment (nslater) [inline devs/docs/logging.rst]: "the logger instance"

File: devs/docs/logging.rst [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+           logger = Loggers.getLogger(MyClass.class, settings);
+           ...
+        }
+  ...
+  }
+
+In your log messages, try to include also information about the "environment"
+in which the operation takes place. Such information can be for example the
+`nodeOperations`, the `phaseId`, or others. Please also consult the guidelines
+presented in logfmt_.
```
Comment (nslater) [inline devs/docs/logging.rst]: > In your log messages, try to include information about the "environment" in which the operation takes place. For example, the `nodeOperations`, `phaseId`, and so on. Please also consult the \`logfmt\`_ guidelines.

File: devs/docs/logging.rst [hunk 8]
```diff
@@ -48,9 +48,8 @@ logger instance should become a class attribute which is initialized with the
   ...
   }
 
-In your log messages, try to include also information about the "environment"
-in which the operation takes place. Such information can be for example the
-`nodeOperations`, the `phaseId`, or others. Please also consult the guidelines
-presented in logfmt_.
+In your log messages, try to include information about the "environment" in
+which the operation takes place. For example, the nodeOperations, phaseId, and
```
Comment (nslater) [inline devs/docs/logging.rst]: please add literal quotes

**Pull request merged**